File: src/main/java/at/meks/calculation/OriginalPowerDataAggregation.java
package at.meks.calculation;

import at.meks.powerdata.PowerData;
import at.meks.powerdata.SinglePowerData;

import java.time.Month;
import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;

public class OriginalPowerDataAggregation {

    private final Map<Month, Double> consumptionFromGrid = new HashMap<>();
    private final Map<Month, Double> fedInToGrid = new HashMap<>();
    private final Collection<Integer> years;

    public OriginalPowerDataAggregation(PowerData powerData, Collection<Integer> years) {
        this.years = years;
        Arrays.stream(Month.values()).forEach(month -> consumptionFromGrid.put(month, 0.0));
        Arrays.stream(Month.values()).forEach(month -> fedInToGrid.put(month, 0.0));
        powerData.stream().forEach(this::calculatePowerStatistics);
    }

    private void calculatePowerStatistics(SinglePowerData singlePowerData) {
        consumptionFromGrid.compute(singlePowerData.timestampUntil().getMonth(),
                                    (month, currentConsumption) ->
                                            currentConsumption == null ? singlePowerData.consumptionKwh() :
                                                    currentConsumption + singlePowerData.consumptionKwh());
        fedInToGrid.compute(singlePowerData.timestampUntil().getMonth(),
                            (month, currentValue) ->
                                    currentValue == null ? singlePowerData.fedInKwh() : currentValue + singlePowerData.fedInKwh());
    }

    public String consumptionPerMonthAsString() {
        return consumptionFromGrid.entrySet().stream()
                                  .sorted(Map.Entry.comparingByKey())
                                  .map(Map.Entry::getValue)
                                  .map(String::valueOf)
                                  .collect(Collectors.joining(", "));
    }

    public String fedInPerMonthAsString() {
        return fedInToGrid.entrySet().stream()
                          .sorted(Map.Entry.comparingByKey())
                          .map(Map.Entry::getValue)
                          .map(value -> value * -1.0)
                          .map(String::valueOf)
                          .collect(Collectors.joining(", "));
    }

    public double consumptionOfYear() {
        return consumptionFromGrid.values().stream().flatMapToDouble(DoubleStream::of).sum();
    }

    public double fedInOfYear() {
        return fedInToGrid.values().stream().flatMapToDouble(DoubleStream::of).sum();
    }

    public Integer[] years() {
        return years.stream().sorted(Comparator.reverseOrder()).toArray(Integer[]::new);
    }
}

File: src/main/kotlin/at/meks/calculation/OriginalPowerDataAggregation.kt
package at.meks.calculation

import at.meks.powerdata.PowerData
import at.meks.powerdata.SinglePowerData
import kotlinx.datetime.Month

class OriginalPowerDataAggregation(val years: List<Int>) {

    private val consumptionFromGrid = mutableMapOf<Month, Double>()
    private val fedInToGrid = mutableMapOf<Month, Double>()

    constructor(powerData: PowerData, years: List<Int>) : this(years) {
        Month.entries.forEach { month -> consumptionFromGrid.put(month, 0.0) }
        Month.entries.forEach { month -> fedInToGrid.put(month, 0.0) }
        powerData.getPowerData().forEach(this::calculatePowerStatistics)
    }

    private fun calculatePowerStatistics(singlePowerData: SinglePowerData) {
        consumptionFromGrid.compute(singlePowerData.timestampUntil.month)
        { month, currentConsumption ->
            if (currentConsumption == null) singlePowerData.consumptionKwh else currentConsumption + singlePowerData.consumptionKwh
        }
        fedInToGrid.compute(singlePowerData.timestampUntil.month)
        { month, currentValue ->
            if (currentValue == null) singlePowerData.fedInKwh else currentValue + singlePowerData.fedInKwh
        }
    }

    fun consumptionPerMonthAsString(): String {
        return consumptionFromGrid.toSortedMap()
            .map { it.value.toString() }
            .joinToString(separator = ", ")
    }

    fun fedInPerMonthAsString(): String {
        return fedInToGrid.toSortedMap()
            .map { it.value * -1.0 }
            .joinToString(separator = ", ") { it.toString() }
    }

    fun consumptionOfYear(): Double {
        return consumptionFromGrid.values.sum()
    }

    fun fedInOfYear(): Double {
        return fedInToGrid.values.sum()
    }

}
